### PR TITLE
FLUID-6223: Add support for EL paths to `fluid.stringTemplate`.

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [FLUID-6223](https://issues.fluidproject.org/browse/FLUID-6223) for details.

### CODE REVIEW

File: src/framework/core/js/Fluid.js [hunk 1]
```diff
@@ -2851,32 +2851,88 @@ var fluid = fluid || fluid_3_0_0;
...
+     *
+     */
+    // unsupported, non-API function
+    fluid.flattenObjectPaths = function (originalObject) {
+        var flattenedObject = {};
+        fluid.each(originalObject, function (value, key) {
+            if (typeof value === "object") {
+                var flattenedSubObject = fluid.flattenObjectPaths(value);
+                fluid.each(flattenedSubObject, function (subValue, subKey) {
+                    flattenedObject[key + "." + subKey] = subValue;
```
Comment (amb26) [inline src/framework/core/js/Fluid.js]: This function should expose configuration which allows the user the option of using the `fluid.model.composePath" algorithm https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/DataBinding.js#L183 (or some other, e.g. the JSON pointer algorithm) as well as this default which simply composes path segments as strings. We may consider this out of scope for now in which case this can just be written up as a JIRA.

File: src/framework/core/js/Fluid.js [hunk 2]
```diff
@@ -2851,32 +2851,88 @@ var fluid = fluid || fluid_3_0_0;
...
+     * }
+     * ```
+     *
+     * This function preserves the previous functionality of displaying an entire object using its `toString` function,
+     * which is why many of the paths above resolve to "[object Object]".
+     *
+     * This function is an unsupported non-API function that is used in by `fluid.stringTemplate` (see below).
+     *
+     * @param `originalObject` `{Object}` - An object.
+     * @returns `{Object}` - A representation of the original object that only contains top-level sub-elements whose keys are EL Paths.
```
Comment (amb26) [inline src/framework/core/js/Fluid.js]: We should try to be consistent with these doc comments in case anyone does write a parser for them. i) I think the general convention is for @return rather than @returns, and ii) it doesn't look like we escape these with backticks

File: src/framework/core/js/Fluid.js [hunk 3]
```diff
@@ -2851,32 +2851,88 @@ var fluid = fluid || fluid_3_0_0;
...
         keys = keys.sort(fluid.compareStringLength());
         for (var i = 0; i < keys.length; ++i) {
             var key = keys[i];
-            var re = fluid.stringToRegExp("%" + key, "g");
-            template = template.replace(re, values[key]);
+            var templatePlaceholder = "%" + key;
+            var replacementValue = flattenedValues[key];
+
+            var indexOfPlaceHolder = -1;
+            while  ((indexOfPlaceHolder = template.indexOf(templatePlaceholder)) !== -1) {
```
Comment (amb26) [inline src/framework/core/js/Fluid.js]: Stray sapce here

File: src/framework/core/js/MessageResolver.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     *
+     * Look up the first matching message template, starting with the current grade and working up through its parents.
+     * Returns both the template for the message and the function used to resolve the localised value.  By default
+     * the resolve function is `fluid.stringTemplate`, and the template returned uses its syntax.
+     *
+     * @param `that` `{Object}` - The component itself.
+     * @param `messagecodes` `{Array}` - One or more message codes to look up templates for.
+     * @returns `{Object}` - An object that contains`template` and `resolveFunc` members (see above).
```
Comment (amb26) [inline src/framework/core/js/MessageResolver.js]: See doc comment comments above

File: src/framework/core/js/Fluid.js (hunk 2 above)
Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: I guess I got the idea that it was OK by omission, i.e. comments [like this one](https://github.com/GPII/gpii-diff/pull/1#pullrequestreview-75075782) that overlooked the backticks.  It's a shame, as I finally trained myself to use them consistently, but I agree, prior art in Infusion doesn't use it (or `returns`).  Will fix shortly.

Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: In reflecting on it further, I think I picked up the habit when adapting JSDocs into markdown.

Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: Agree on the goal and the specifics.  I notice a few other inconsistencies within the package, for example, some JSDocs use `@param variableName {Type}`, others use `@param {Type} variableName`.  I propose standardising on the former.

Comment (amb26) [reply src/framework/core/js/Fluid.js]: Yes, that seems good - feel free to normalise any examples of the other that you find

Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: Looks like we can at least enforce the consistent use of `@returns`:  https://eslint.org/docs/rules/valid-jsdoc#options

If you're game, I can submit a PR against our comment ESLint rules to add that.

File: src/framework/core/js/Fluid.js (hunk 3 above)
Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: Wow, good eye!

File: src/framework/core/js/Fluid.js (hunk 2 above)
Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: Anyway, fixed throughout.

File: src/framework/core/js/MessageResolver.js (hunk 4 above)
Comment (the-t-in-rtf) [reply src/framework/core/js/MessageResolver.js]: Done.

File: src/framework/core/js/Fluid.js (hunk 1 above)
Comment (the-t-in-rtf) [reply src/framework/core/js/Fluid.js]: I wrote up a ticket for future work in this area:  https://issues.fluidproject.org/browse/FLUID-6243

**Comment by the-t-in-rtf** [comment]
@amb26, I have addressed this round of feedback, this should be ready for review again.

**Comment by the-t-in-rtf** [comment]
@amb26  and I discussed a few more oddities in the JSDocs via chat that have now been addressed.  This is ready for review again.

**Comment by the-t-in-rtf** [comment]
@amb26, I need you to cut a dev release of this so I can exercise it in my work with the i18n spike against gpii-handlebars.

File: tests/framework-tests/core/js/FluidJSTests.js [hunk 5]
```diff
@@ -391,6 +391,61 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
...
+        jqUnit.assertDeepEq("A top-level array should be flattened correctly.", { 0: "monkey", 1: "fighting", 2: "snakes"}, output);
+    });
+
+    jqUnit.test("flattenObjectPaths: deep array", function () {
+        var originalObject = { "plane": ["monday", "to", "friday"]};
+        var output = fluid.flattenObjectPaths(originalObject);
+        jqUnit.assertDeepEq("A deep array should be flattened correctly.", { "plane": "monday,to,friday", "plane.0": "monday", "plane.1": "to", "plane.2": "friday"}, output);
+    });
+
+    jqUnit.test("flattenObjectPaths: deep empty objects", function () {
```
Comment (amb26) [inline tests/framework-tests/core/js/FluidJSTests.js]: We should add a test to verify that the code doesn't explode with a `null` value as a member. I think, by eye, that it won't, but because it features a typeof === "object" check there is the suggestion of a risk

File: tests/framework-tests/core/js/FluidJSTests.js [hunk 6]
```diff
@@ -391,6 +391,61 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
...
+                    value: "foo"
+                }
+            },
+            otherPath: {
+                to: {
+                    otherValue: "bar"
+                }
+            }
+        };
+        var output         = fluid.flattenObjectPaths(originalObject);
```
Comment (amb26) [inline tests/framework-tests/core/js/FluidJSTests.js]: Stray spaces of that peculiar nature
